csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -199,31 +440,108 @@ std::pair<int64_t, int64_t> getPersistentBufferSize(
-            persistent_buffer_size_info.projected_persistent_buffer_size);
+  // Project to inputs when there is at least one outer broadcast tensor or
+  // projected persistent buffer size is smaller. When projecting to inputs, the
+  // outer broadcast tensor is reused in the loop over the iteration dimension,
+  // test shows it is faster than the non-projected version which requires
+  // reload from gmem for each iteration.
+  // Note: in current use cases (layer norm bwd and RMS norm bwd), there are
+  // outer broadcast tvs and always project to inputs.
+  const auto& outer_broadcast_tvs = getOuterBroadcastTvs(fusion, reduction_tvs);
+  buffer_params.project_to_input =
+      normalization_scheduler_utils::isProjectBufferToInputs(
+          fusion,
+          runtime_info,
+          persistent_buffer_info,
+          persistent_buffer_size_info,
+          ScheduleHeuristic::InnerOuterPersistent,
+          outer_broadcast_tvs.empty());
 
-  // in combined_inner_outer_reduction, the partial results of outer
-  // reductions must be persistent, allow register spill avoid segmentation
-  std::vector<TensorView*> outer_reduction_tvs;
-  for (auto tv : reduction_tvs) {
-    if (!scheduler_utils::isFastestDimReduction(tv)) {
-      outer_reduction_tvs.emplace_back(tv);
+  auto total_buffer_size = buffer_params.project_to_input
+      ? persistent_buffer_size_info.projected_persistent_buffer_size
+      : persistent_buffer_size_info.persistent_buffer_size;
+  total_buffer_size +=
+      partialOuterReductionBufferSize(reduction_tvs, runtime_info);
+
+  const auto dev_prop = at::cuda::getCurrentDeviceProperties();
+  int64_t smem_overhead = scheduler_utils::getSharedMemoryOverheadPerBlock(
+      fusion,
+      reduction_tvs,
+      InnerOuterPersistentKernelScheduler::threads_per_block_max);
+  int64_t available_smem =
+      (int64_t)dev_prop->sharedMemPerMultiprocessor - smem_overhead;
+  int64_t available_regs = scheduler_utils::register_file_size_full;
+  buffer_params.smem_overhead = smem_overhead;
+
+  // Put all the persistent tensors in registers
+  buffer_params.regs_buffer_size = total_buffer_size;
+  buffer_params.smem_buffer_size = 0;
+
+  // Relocate buffers to shared memory until the buffer size in registers is
+  // within the allowable limit.
+  if (buffer_params.regs_buffer_size > available_regs) {
+    const auto sorted_candidate_tvs = buffer_params.project_to_input
+        ? sortProjectableBufferInputs(
+              fusion,
+              persistent_buffer_info.persistent_buffers,
+              persistent_buffer_info.projectable_buffer_inputs,
+              outer_broadcast_tvs,
+              reduction_tvs,
+              runtime_info)
+        : sortPersistentBuffers(persistent_buffer_info.persistent_buffers);
+
+    // determine the number of buffers to move to shared memory
+    const int64_t n_buffers = (int64_t)sorted_candidate_tvs.size();
+    int64_t acc_regs_buffer_size = 0, acc_smem_buffer_size = 0;
```

Comment:
Please comment these variables

Reply:
` acc_regs_buffer_size and acc_smem_buffer_size` are removed.